src/Microsoft.AspNetCore.Mvc.Core/Internal/FileContentResultExecutor.cs

```diff
@@ -15,15 +17,51 @@ public FileContentResultExecutor(ILoggerFactory loggerFactory)
 
         public Task ExecuteAsync(ActionContext context, FileContentResult result)
         {
-            SetHeadersAndLog(context, result);
-            return WriteFileAsync(context, result);
+            var rangeInfo = new(RangeItemHeaderValue range, long rangeLength)?();
+            if (result.LastModified.HasValue)
+            {
+                rangeInfo = SetHeadersAndLog(
+                    context,
+                    result,
+                    result.FileContents.Length,
+                    result.LastModified.Value,
+                    result.EntityTag);
+            }
+            else
+            {
+                rangeInfo = SetHeadersAndLog(
+                    context,
+                    result,
+                    result.FileContents.Length);
+            }
+
+            if (rangeInfo.HasValue)
+            {
+                return WriteFileAsync(context, result, rangeInfo.Value.range, rangeInfo.Value.rangeLength);
+            }
+
+            return WriteFileAsync(context, result, null, 0);
         }
 
-        private static Task WriteFileAsync(ActionContext context, FileContentResult result)
+        private static Task WriteFileAsync(ActionContext context, FileContentResult result, RangeItemHeaderValue range, long rangeLength)
         {
             var response = context.HttpContext.Response;
+            var outputStream = response.Body;
+
+            if (range == null)
+            {
+                return response.Body.WriteAsync(result.FileContents, offset: 0, count: result.FileContents.Length);
+            }
+
+            else if (rangeLength == 0)
+            {
+                return Task.CompletedTask;
+            }
 
-            return response.Body.WriteAsync(result.FileContents, offset: 0, count: result.FileContents.Length);
+            else
+            {
+                return response.Body.WriteAsync(result.FileContents, offset: (int)range.From.Value, count: (int)rangeLength);
```

Comment:
> This casting from long to int looks dangerous

I don't like it either. All the other file types use `StreamCopyOperation` [here](https://github.com/aspnet/Mvc/blob/jbagga/ByteRange3702/src/Microsoft.AspNetCore.Mvc.Core/Internal/FileStreamResultExecutor.cs#L74)